conans/client/cmd/uploader.py

```diff
@@ -92,62 +94,90 @@ def _upload(self, conan_file, conan_ref, packages_ids, retry, retry_wait,
             remote_manifest = None
 
         self._user_io.out.info("Uploading %s to remote '%s'" % (str(conan_ref), recipe_remote.name))
-        self._upload_recipe(conan_ref, retry, retry_wait, policy, recipe_remote, remote_manifest)
-
-        if not cur_recipe_remote and policy != UPLOAD_POLICY_SKIP:
-            self._registry.refs.set(conan_ref, recipe_remote.name)
 
-        recorder.add_recipe(conan_ref, recipe_remote.name, recipe_remote.url)
+        metadata = self._client_cache.load_metadata(conan_ref)
+        ref = conan_ref.copy_with_rev(metadata.recipe.revision)
+        self._upload_recipe(ref, retry, retry_wait, policy, recipe_remote, remote_manifest)
 
+        recorder.add_recipe(ref, recipe_remote.name, recipe_remote.url)
         if packages_ids:
+            # Filter packages that don't match the recipe revision
+            revisions_enabled = get_env("CONAN_CLIENT_REVISIONS_ENABLED", False)
+            if revisions_enabled and ref.revision:
+                recipe_package_ids = []
+                for package_id in packages_ids:
+                    pref = PackageReference(ref, package_id)
+                    metadata = self._client_cache.load_metadata(pref.conan)
```

Comment:
This metadata has already been loaded above, if it is using ``ref`` for it, it should be the same metadata, shouldn't it?